ttnn/cpp/ttnn/operations/data_movement/slice/slice.cpp

```diff
@@ -8,133 +8,239 @@
+    std::array<uint32_t, 4> padded_ends = modified_ends;
+    if (input_tensor.layout() == Layout::TILE) {
+        padded_ends[2] = detail::round_up_to_multiple_of_32(padded_ends[2]);
+        padded_ends[3] = detail::round_up_to_multiple_of_32(padded_ends[3]);
+    }
+    std::vector<uint32_t> actual_shape, padded_shape;
+    actual_shape.reserve(input_rank);
+    padded_shape.reserve(input_rank);
+    bool empty = false;
+    for (int i = 0; i < input_rank; ++i) {
+        // Check that end indices are greater than or equal to start indices (empty tensor where end=start is supported)
+        TT_FATAL(modified_ends[i + rank_diff] >= modified_begins[i + rank_diff], "End must be greater than or equal to start");
+        auto val = output_dim_i(i + rank_diff, modified_ends);
+        if (val == 0) {
+            empty = true;
         }
+        actual_shape.push_back(val);
+        padded_shape.push_back(std::max(output_dim_i(i + rank_diff, padded_ends), (uint32_t)1));
     }
+
+    ttnn::Shape output_shape(actual_shape, padded_shape);
+    // PyTorch supports final dimension = 0 (start = end, where end is inclusive) so >= is okay, just return an empty tensor
+    if (empty) {
+        return ttnn::empty(output_shape, input_tensor.dtype(), input_tensor.layout(),
+            input_tensor.device(), memory_config_arg.value_or(input_tensor.memory_config()));
+    }
+
+    // Early exit if slice is a no-op (ends = padding ends and step = 1 for all dimensions)
+    bool no_step = std::all_of(step.begin(), step.end(), [](int i) {return i == 1;});
+    if (tt::tt_metal::LegacyShape(padded_shape) == input_tensor.get_legacy_shape() and no_step) {
+        return input_tensor;
+    }
+
     if (input_tensor.storage_type() != StorageType::DEVICE) {
-        TT_FATAL(!modified_step.has_value(), "Host tensor slice does not support strides");
-        tt::tt_metal::LegacyShape output_tensor_shape = {
-            output_tensor_end[0] - output_tensor_start[0] + 1,
-            output_tensor_end[1] - output_tensor_start[1] + 1,
-            output_tensor_end[2] - output_tensor_start[2] + 1,
-            output_tensor_end[3] - output_tensor_start[3] + 1,
-        };
+        TT_FATAL(no_step, "Host tensor slice does not support strides");
         // if we support negative strides, we can't do this early exit
-        if (input_tensor.get_legacy_shape() == output_tensor_shape) {
+        if (input_tensor.get_legacy_shape() == actual_shape) {
             return input_tensor;
         } else {
-            return input_tensor.unpad(output_tensor_start, output_tensor_end);
+            auto input_4d_rm = ttnn::to_layout(input_4d, Layout::ROW_MAJOR, std::nullopt, std::nullopt, (Device *)nullptr);
+            auto output_4d =  input_4d_rm.unpad(tt::tt_metal::LegacyShape(modified_begins), tt::tt_metal::LegacyShape(modified_ends));
+            auto output_4d_rm = ttnn::to_layout(output_4d, input_tensor.get_layout(), std::nullopt, std::nullopt, (Device *)nullptr);
```

Comment:
I don't fully follow whats going on here. Can you clarify please?

Reply:
I'm converting to ROW Major before doing the host-side unpad. This is mimicking the logic in the python-side __getItem__ overload, which I assume is there because of unpad's constraints.